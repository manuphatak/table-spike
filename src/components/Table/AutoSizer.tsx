import React, { ReactElement, useCallback, useState } from "react"
import Measure, { BoundingRect, ContentRect } from "react-measure"
import styled from "styled-components"

type Dimensions = Pick<BoundingRect, "height" | "width">

interface AutoSizerProps {
  children: (props: { dimensions: Dimensions }) => React.ReactNode
}

const Container = styled.div`
  width: 100%;
  height: 100%;
`

export default function AutoSizer(props: AutoSizerProps): ReactElement {
  const [dimensions, setDimensions] = useState<Dimensions>({
    width: -1,
    height: -1,
  })

  const handleResize = useCallback((contentRect: ContentRect) => {
    if (contentRect.bounds) {
      setDimensions(contentRect.bounds)
    }
  }, [])

  return (
    <Measure bounds onResize={handleResize}>
      {({ measureRef }) => (
        <Container ref={measureRef}>{props.children({ dimensions })}</Container>
      )}
    </Measure>
  )
}
