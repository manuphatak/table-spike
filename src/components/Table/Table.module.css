.List {
  border: 1px solid #d9dddd;
  font-family: sans-serif;
  font-size: 12px;
  box-sizing: border-box;
}

.ListItemEven,
.ListItemOdd {
  display: flex;
  align-items: center;
  justify-content: center;
}

.ListItemEven {
  background-color: #f8f8f0;
}
