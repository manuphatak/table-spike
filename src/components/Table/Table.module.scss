@import "~@procore/core-icons";
@import "~@procore/core-css/src/components/_tables.scss";
@import "~@procore/core-css/src/components/_icon.scss";
