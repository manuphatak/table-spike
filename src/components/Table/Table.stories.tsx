import { optionsKnob, withKnobs } from "@storybook/addon-knobs"
import { prop } from "ramda"
import React, { useState } from "react"
import CAR_DATA from "../../fixtures/CAR_DATA"
import Table from "./"
import {
  ColumnDefinition,
  GroupDefinition,
  RowDefinitions,
  RowType,
  generateGroups,
  DispatchHandler,
  DispatchEvent,
} from "./Table"

export default {
  title: "Table",
  component: Table,
  decorators: [withKnobs],
}
type ArrayInfer<T extends any[]> = T extends Array<infer U> ? U : never
type CarDatum = ArrayInfer<typeof CAR_DATA>

const columnDefinitions: ColumnDefinition<CarDatum>[] = [
  { dataKey: "car_make", label: "Car Make", flex: "1 0 100px" },
  { dataKey: "car_model", label: "Car Model", flex: "1 0 100px" },
  { dataKey: "car_year", label: "Car Year", flex: "0.5 0 50px" },
  { dataKey: "country", label: "Country", flex: "1 0 100px" },
  { dataKey: "car_price", label: "Price", flex: "1 0 100px" },
  { dataKey: "comments", label: "Comments", flex: "2 0 200px" },
]

const rowDefinitions: RowDefinitions<CarDatum> = {
  [RowType.Body]: {
    height: 48,
    keyFn: prop("id"),
  },
}

export const Basic = () => {
  const groupDefinitions = groupDefinitionsKnob()
  const groups = generateGroups<CarDatum>(
    rowDefinitions,
    groupDefinitions,
    CAR_DATA
  )

  const [collapsedGroupPaths, setCollapsedGroupPaths] = useState([
    JSON.stringify(["China"]),
    JSON.stringify(["Japan", "Ford"]),
  ])

  const dispatch: DispatchHandler = (action) => {
    switch (action.type) {
      case DispatchEvent.OnCollapse:
        ;(() => {
          if (action.collapsed) {
            setCollapsedGroupPaths([...collapsedGroupPaths, action.path])
          } else {
            setCollapsedGroupPaths(
              collapsedGroupPaths.filter((path) => path !== action.path)
            )
          }
        })()
    }
  }
  return (
    <Table
      data={CAR_DATA}
      groups={groups}
      columnDefinitions={columnDefinitions}
      collapsedGroupPaths={collapsedGroupPaths}
      dispatch={dispatch}
    />
  )
}

function groupDefinitionsKnob(): GroupDefinition<CarDatum>[] {
  const options = Object.fromEntries(
    columnDefinitions.map(({ label, dataKey }) => [label, dataKey])
  )

  const groupValues = (optionsKnob("Groups", options, ["country", "car_make"], {
    display: "multi-select",
  }) as unknown) as (keyof CarDatum)[] | null

  if (!groupValues) {
    return []
  }
  return groupValues.map((value) => (data) => prop(value, data).toString())
}
