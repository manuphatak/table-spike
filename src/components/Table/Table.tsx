import classnames from "classnames/bind"
import React, { CSSProperties, ReactElement } from "react"
import { VariableSizeList } from "react-window"
import styled from "styled-components/macro"
import AutoSizer from "./AutoSizer"
import carData from "./carData"
import styles from "./Table.module.scss"

const cx = classnames.bind(styles)

interface TableProps {}

const VirtualList = styled(VariableSizeList)`
  box-sizing: border-box;
`

const Row = styled.div`
  display: flex;
  flex-flow: row nowrap;
`

const RowColumn = styled.div`
  flex: 1 0 100px;
  white-space: nowrap;
  overflow: hidden;
`

const renderableCarData = carData.map((car) => ({
  ...car,
  height: 48,
}))

// const columnDefinitions = [
//   { dataKey: "car_make", label: "Car Make" },
//   { dataKey: "car_model", label: "Car Model" },
//   { dataKey: "car_year", label: "Car Year" },
//   { dataKey: "country", label: "Country" },
//   { dataKey: "car_price", label: "Price" },
//   { dataKey: "comments", label: "Comments" },
// ]

const renderRow = (props: {
  index: number
  style: CSSProperties
}): JSX.Element => {
  const rowData = renderableCarData[props.index]
  return (
    <Row style={props.style} className={cx("table__body-row")}>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.car_make}</label>
      </RowColumn>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.car_model}</label>
      </RowColumn>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.car_year}</label>
      </RowColumn>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.country}</label>
      </RowColumn>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.car_price}</label>
      </RowColumn>
      <RowColumn className={cx("table__td")}>
        <label className={cx("table__text-cell")}>{rowData.comments}</label>
      </RowColumn>
    </Row>
  )
}

const getItemSize = (index: number): number => renderableCarData[index].height

export default function Table(_props: TableProps): ReactElement {
  return (
    <AutoSizer>
      {({ dimensions }) => (
        <VirtualList
          height={dimensions.height}
          itemCount={carData.length}
          itemSize={getItemSize}
          width={dimensions.width}
        >
          {renderRow}
        </VirtualList>
      )}
    </AutoSizer>
  )
}
