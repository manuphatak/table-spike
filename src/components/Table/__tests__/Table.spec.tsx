import { render } from "enzyme"
import { prop } from "ramda"
import React from "react"
import Table, {
  addCommonRowData,
  createGroups,
  flattenGroups,
  RowType,
  ColumnDefinition,
  RowDefinitions,
  GroupDefinition,
} from "../Table"
import CAR_DATA from "../../../fixtures/CAR_DATA"

type ArrayInfer<T extends any[]> = T extends Array<infer U> ? U : never
type CarDatum = ArrayInfer<typeof CAR_DATA>

const columnDefinitions: ColumnDefinition<CarDatum>[] = [
  { dataKey: "car_make", label: "Car Make", flex: "1 0 100px" },
  { dataKey: "car_model", label: "Car Model", flex: "1 0 100px" },
  { dataKey: "car_year", label: "Car Year", flex: "0.5 0 50px" },
  { dataKey: "country", label: "Country", flex: "1 0 100px" },
  { dataKey: "car_price", label: "Price", flex: "1 0 100px" },
  { dataKey: "comments", label: "Comments", flex: "2 0 200px" },
]

const rowDefinitions: RowDefinitions<CarDatum> = {
  [RowType.Body]: {
    height: 48,
    keyFn: prop("id"),
  },
}

const groupDefinitions: GroupDefinition<CarDatum>[] = [prop("car_make")]

describe("render", () => {
  it("matches snapshot", () => {
    const wrapper = render(
      <Table
        data={CAR_DATA}
        initialDimensions={{
          height: 400,
          width: 1600,
        }}
        columnDefinitions={columnDefinitions}
        rowDefinitions={rowDefinitions}
        groupDefinitions={groupDefinitions}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})

describe("createGroups", () => {
  const data = addCommonRowData(rowDefinitions, [
    {
      id: 3,
      country: "Kazakhstan",
      car_make: "Chevrolet",
      car_model: "Camaro",
      car_year: 1976,
      car_price: "$5602.92",
      comments: "e-enable synergistic e-markets",
    },
    {
      id: 11,
      country: "Brazil",
      car_make: "Cadillac",
      car_model: "Escalade EXT",
      car_year: 2006,
      car_price: "$7834.39",
      comments: "drive interactive systems",
    },
    {
      id: 12,
      country: "Philippines",
      car_make: "Ford",
      car_model: "Model T",
      car_year: 1909,
      car_price: "$8869.46",
      comments: "redefine seamless paradigms",
    },
    {
      id: 25,
      country: "Czech Republic",
      car_make: "Cadillac",
      car_model: "SRX",
      car_year: 2007,
      car_price: "$6722.29",
      comments: "repurpose front-end interfaces",
    },
    {
      id: 101,
      country: "Russia",
      car_make: "Cadillac",
      car_model: "Escalade EXT",
      car_year: 2002,
      car_price: "$7713.56",
      comments: "incentivize bricks-and-clicks deliverables",
    },
  ])

  describe("given an empty array of group functions", () => {
    it("is a noop", () => {
      expect(createGroups([], data)).toEqual(data)
    })
  })

  describe("given an array of group functions", () => {
    it("groups data", () => {
      expect(createGroups([prop("car_make")], data)).toEqual([
        {
          value: "Chevrolet",
          path: ["Chevrolet"],
          children: addCommonRowData(rowDefinitions, [
            {
              id: 3,
              country: "Kazakhstan",
              car_make: "Chevrolet",
              car_model: "Camaro",
              car_year: 1976,
              car_price: "$5602.92",
              comments: "e-enable synergistic e-markets",
            },
          ]),
        },
        {
          value: "Cadillac",
          path: ["Cadillac"],
          children: addCommonRowData(rowDefinitions, [
            {
              id: 11,
              country: "Brazil",
              car_make: "Cadillac",
              car_model: "Escalade EXT",
              car_year: 2006,
              car_price: "$7834.39",
              comments: "drive interactive systems",
            },
            {
              id: 25,
              country: "Czech Republic",
              car_make: "Cadillac",
              car_model: "SRX",
              car_year: 2007,
              car_price: "$6722.29",
              comments: "repurpose front-end interfaces",
            },
            {
              id: 101,
              country: "Russia",
              car_make: "Cadillac",
              car_model: "Escalade EXT",
              car_year: 2002,
              car_price: "$7713.56",
              comments: "incentivize bricks-and-clicks deliverables",
            },
          ]),
        },
        {
          value: "Ford",
          path: ["Ford"],
          children: addCommonRowData(rowDefinitions, [
            {
              id: 12,
              country: "Philippines",
              car_make: "Ford",
              car_model: "Model T",
              car_year: 1909,
              car_price: "$8869.46",
              comments: "redefine seamless paradigms",
            },
          ]),
        },
      ])
    })

    it("recursively groups data", () => {
      expect(createGroups([prop("car_make"), prop("car_model")], data)).toEqual(
        [
          {
            value: "Chevrolet",
            path: ["Chevrolet"],
            children: [
              {
                value: "Camaro",
                path: ["Chevrolet", "Camaro"],
                children: addCommonRowData(rowDefinitions, [
                  {
                    id: 3,
                    country: "Kazakhstan",
                    car_make: "Chevrolet",
                    car_model: "Camaro",
                    car_year: 1976,
                    car_price: "$5602.92",
                    comments: "e-enable synergistic e-markets",
                  },
                ]),
              },
            ],
          },
          {
            value: "Cadillac",
            path: ["Cadillac"],
            children: [
              {
                value: "Escalade EXT",
                path: ["Cadillac", "Escalade EXT"],
                children: addCommonRowData(rowDefinitions, [
                  {
                    id: 11,
                    country: "Brazil",
                    car_make: "Cadillac",
                    car_model: "Escalade EXT",
                    car_year: 2006,
                    car_price: "$7834.39",
                    comments: "drive interactive systems",
                  },
                  {
                    id: 101,
                    country: "Russia",
                    car_make: "Cadillac",
                    car_model: "Escalade EXT",
                    car_year: 2002,
                    car_price: "$7713.56",
                    comments: "incentivize bricks-and-clicks deliverables",
                  },
                ]),
              },
              {
                value: "SRX",
                path: ["Cadillac", "SRX"],
                children: addCommonRowData(rowDefinitions, [
                  {
                    id: 25,
                    country: "Czech Republic",
                    car_make: "Cadillac",
                    car_model: "SRX",
                    car_year: 2007,
                    car_price: "$6722.29",
                    comments: "repurpose front-end interfaces",
                  },
                ]),
              },
            ],
          },
          {
            value: "Ford",
            path: ["Ford"],
            children: [
              {
                value: "Model T",
                path: ["Ford", "Model T"],
                children: addCommonRowData(rowDefinitions, [
                  {
                    id: 12,
                    country: "Philippines",
                    car_make: "Ford",
                    car_model: "Model T",
                    car_year: 1909,
                    car_price: "$8869.46",
                    comments: "redefine seamless paradigms",
                  },
                ]),
              },
            ],
          },
        ]
      )
    })
  })
})

describe("flattenGroups", () => {
  const data = addCommonRowData(rowDefinitions, [
    {
      id: 3,
      country: "Kazakhstan",
      car_make: "Chevrolet",
      car_model: "Camaro",
      car_year: 1976,
      car_price: "$5602.92",
      comments: "e-enable synergistic e-markets",
    },
    {
      id: 11,
      country: "Brazil",
      car_make: "Cadillac",
      car_model: "Escalade EXT",
      car_year: 2006,
      car_price: "$7834.39",
      comments: "drive interactive systems",
    },
    {
      id: 12,
      country: "Philippines",
      car_make: "Ford",
      car_model: "Model T",
      car_year: 1909,
      car_price: "$8869.46",
      comments: "redefine seamless paradigms",
    },
    {
      id: 25,
      country: "Czech Republic",
      car_make: "Cadillac",
      car_model: "SRX",
      car_year: 2007,
      car_price: "$6722.29",
      comments: "repurpose front-end interfaces",
    },
    {
      id: 101,
      country: "Russia",
      car_make: "Cadillac",
      car_model: "Escalade EXT",
      car_year: 2002,
      car_price: "$7713.56",
      comments: "incentivize bricks-and-clicks deliverables",
    },
  ])

  describe("given flat data", () => {
    it("is a noop", () => {
      expect(flattenGroups(data)).toEqual(data)
    })
  })

  describe("given a single grouping fn", () => {
    it("flattens the group into rows", () => {
      expect(flattenGroups(createGroups([prop("car_make")], data))).toEqual([
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Chevrolet"]),
          label: "Chevrolet",
          depth: 0,
        },
        {
          type: RowType.Body,
          id: 3,
          country: "Kazakhstan",
          car_make: "Chevrolet",
          car_model: "Camaro",
          car_year: 1976,
          car_price: "$5602.92",
          comments: "e-enable synergistic e-markets",
          height: 48,
          key: 3,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Cadillac"]),
          label: "Cadillac",
          depth: 0,
        },
        {
          type: RowType.Body,
          id: 11,
          country: "Brazil",
          car_make: "Cadillac",
          car_model: "Escalade EXT",
          car_year: 2006,
          car_price: "$7834.39",
          comments: "drive interactive systems",
          height: 48,
          key: 11,
        },
        {
          type: RowType.Body,
          id: 25,
          country: "Czech Republic",
          car_make: "Cadillac",
          car_model: "SRX",
          car_year: 2007,
          car_price: "$6722.29",
          comments: "repurpose front-end interfaces",
          height: 48,
          key: 25,
        },
        {
          type: RowType.Body,
          id: 101,
          country: "Russia",
          car_make: "Cadillac",
          car_model: "Escalade EXT",
          car_year: 2002,
          car_price: "$7713.56",
          comments: "incentivize bricks-and-clicks deliverables",
          height: 48,
          key: 101,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Ford"]),
          label: "Ford",
          depth: 0,
        },
        {
          type: RowType.Body,
          id: 12,
          country: "Philippines",
          car_make: "Ford",
          car_model: "Model T",
          car_year: 1909,
          car_price: "$8869.46",
          comments: "redefine seamless paradigms",
          height: 48,
          key: 12,
        },
      ])
    })
  })
  describe("given multiple grouping fns", () => {
    it("flattens the group into rows", () => {
      expect(
        flattenGroups(createGroups([prop("car_make"), prop("car_model")], data))
      ).toEqual([
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Chevrolet"]),
          label: "Chevrolet",
          depth: 0,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Chevrolet", "Camaro"]),
          label: "Camaro",
          depth: 1,
        },
        {
          type: RowType.Body,
          id: 3,
          country: "Kazakhstan",
          car_make: "Chevrolet",
          car_model: "Camaro",
          car_year: 1976,
          car_price: "$5602.92",
          comments: "e-enable synergistic e-markets",
          height: 48,
          key: 3,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Cadillac"]),
          label: "Cadillac",
          depth: 0,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Cadillac", "Escalade EXT"]),
          label: "Escalade EXT",
          depth: 1,
        },
        {
          type: RowType.Body,
          id: 11,
          country: "Brazil",
          car_make: "Cadillac",
          car_model: "Escalade EXT",
          car_year: 2006,
          car_price: "$7834.39",
          comments: "drive interactive systems",
          height: 48,
          key: 11,
        },
        {
          type: RowType.Body,
          id: 101,
          country: "Russia",
          car_make: "Cadillac",
          car_model: "Escalade EXT",
          car_year: 2002,
          car_price: "$7713.56",
          comments: "incentivize bricks-and-clicks deliverables",
          height: 48,
          key: 101,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Cadillac", "SRX"]),
          label: "SRX",
          depth: 1,
        },
        {
          type: RowType.Body,
          id: 25,
          country: "Czech Republic",
          car_make: "Cadillac",
          car_model: "SRX",
          car_year: 2007,
          car_price: "$6722.29",
          comments: "repurpose front-end interfaces",
          height: 48,
          key: 25,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Ford"]),
          label: "Ford",
          depth: 0,
        },
        {
          type: RowType.GroupHeader,
          height: 48,
          key: JSON.stringify(["Ford", "Model T"]),
          label: "Model T",
          depth: 1,
        },
        {
          type: RowType.Body,
          id: 12,
          country: "Philippines",
          car_make: "Ford",
          car_model: "Model T",
          car_year: 1909,
          car_price: "$8869.46",
          comments: "redefine seamless paradigms",
          height: 48,
          key: 12,
        },
      ])
    })
  })
})
