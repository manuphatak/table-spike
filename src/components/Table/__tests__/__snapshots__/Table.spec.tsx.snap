// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render matches snapshot 1`] = `
.c0 {
  width: 100%;
  height: 100%;
}

.c1 {
  box-sizing: border-box;
}

.c1 *,
.c1 *:before,
.c1 *:after {
  box-sizing: inherit;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  margin: 0;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  -webkit-flex: 1 0 100px;
  -ms-flex: 1 0 100px;
  flex: 1 0 100px;
  white-space: nowrap;
  overflow: hidden;
}

.c4 {
  -webkit-flex: 0.5 0 50px;
  -ms-flex: 0.5 0 50px;
  flex: 0.5 0 50px;
  white-space: nowrap;
  overflow: hidden;
}

.c5 {
  -webkit-flex: 2 0 200px;
  -ms-flex: 2 0 200px;
  flex: 2 0 200px;
  white-space: nowrap;
  overflow: hidden;
}

<div
  class="c0"
>
  <div
    class="c1 table__table"
    style="position:relative;height:400px;width:1600px;overflow:auto;-webkit-overflow-scrolling:touch;will-change:transform;direction:ltr"
  >
    <div
      style="height:53128px;width:100%"
    >
      <div
        class="c2 table__head"
        style="position:absolute;left:0;top:0;height:48px;width:100%"
      >
        <div
          class="c3 table__th"
        >
          <label
            class="table__th__label"
          >
            Car Make
          </label>
        </div>
        <div
          class="c3 table__th"
        >
          <label
            class="table__th__label"
          >
            Car Model
          </label>
        </div>
        <div
          class="c4 table__th"
        >
          <label
            class="table__th__label"
          >
            Car Year
          </label>
        </div>
        <div
          class="c3 table__th"
        >
          <label
            class="table__th__label"
          >
            Country
          </label>
        </div>
        <div
          class="c3 table__th"
        >
          <label
            class="table__th__label"
          >
            Price
          </label>
        </div>
        <div
          class="c5 table__th"
        >
          <label
            class="table__th__label"
          >
            Comments
          </label>
        </div>
      </div>
      <div
        class="c2 table__group-row table__group-row__content"
        style="position:absolute;left:0;top:48px;height:48px;width:100%"
      >
        <div
          class="table__group-row__title"
        >
          Mercury
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:96px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mountaineer
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            1997
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            China
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $8545.87
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            engage cutting-edge systems
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:144px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Grand Marquis
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            1985
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Indonesia
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $3382.45
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            disintermediate impactful e-business
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:192px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mountaineer
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            2001
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            China
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $3146.50
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            incentivize world-class web-readiness
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:240px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Monterey
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            2005
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            China
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $3586.33
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            extend killer ROI
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:288px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Capri
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            1992
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            China
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $6812.68
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            harness transparent users
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:336px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Sable
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            2002
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Indonesia
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $8796.52
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            envisioneer integrated channels
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:384px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Lynx
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            1984
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Indonesia
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $6908.80
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            incubate viral action-items
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:432px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mariner
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            2009
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Colombia
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $1036.71
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            embrace proactive channels
          </label>
        </div>
      </div>
      <div
        class="c2 table__body-row"
        style="position:absolute;left:0;top:480px;height:48px;width:100%"
      >
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Mercury
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Sable
          </label>
        </div>
        <div
          class="c4 table__td"
        >
          <label
            class="table__text-cell"
          >
            1998
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            Philippines
          </label>
        </div>
        <div
          class="c3 table__td"
        >
          <label
            class="table__text-cell"
          >
            $9516.27
          </label>
        </div>
        <div
          class="c5 table__td"
        >
          <label
            class="table__text-cell"
          >
            architect mission-critical schemas
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
